doc/adr/0002-nats-typed-messages.md

```diff
@@ -0,0 +1,182 @@
+   Cloud Events. Typically, published unsolicited.
+
+Today NATS Server do not support publishing Cloud Events natively however a bridge can be created to publish
+those to other cloud systems using the `jsm.go` package that supports converting `events` into Cloud Event format.
+
+### Message Types
+
+There is no standard way to indicate the schema of a specific message. We looked at a lot of prior art from CNCF
+projects, public clouds and more but found very little commonality. The nearest standard is the Uniform Resource Name
+which still leaves most of the details up to the project and does not conventionally support versioning.
+
+We chose a message type like `io.nats.jetstream.api.v1.consumer_delete_response`, `io.nats.server.advisory.v1.client_connect`
+or `io.nats.unknown_message`.
+
+`io.nats.unknown_message` is a special type returned for anything without valid type hints. In go that implies
+`map[string]interface{}`.
+
+The structure is as follows: io.nats.`<source>`.`<catagory>`.v`<version>`.`<name>`
+
+#### Source
+
+The project is the overall originator of a message and should be short but descriptive, today we have 2 - `server` and `
+jetstream` - as we continue to build systems around Stream Processing and more we'd add more of these types. I anticipate
+for example adding a few to Surveyor for publishing significant lifecycle events.
+
+Generated Cloud Events messages has the `source` set to `urn:nats:<source>`.
+
+|Project|Description|
+|-------|-----------|
+|`server`|The core NATS Server excluding JetStream related messages|
+|`jetstream`|Any JetStream related message|
+
+#### Category
+
+The `category` groups messages by related sub-groups of the `source`, often this also appears in the subjects
+these messages get published to.
+
+This is a bit undefined, examples in use now are `api`, `advisory`, `metric`. Where possible try to fit in with
+existing chosen ones, if none suits update this table with your choice and try to pick generic category names.
+
+|Category|Description|
+|----|-----------|
+|`api`|Typically these are `messages` used in synchronous request response APIs|
+|`advisory`|These are `events` that describe a significant event that happened like a client connecting or disconnecting|
+|`metric`|These are `events` that relate to monitoring - how long did it take a message to be acknowledged|
+
+#### Versioning
+
+The ideal outcome is that we never need to version any message and maintain future compatibility. 
+
+We think we can do that with the JetStream API. Monitoring, Observability and black box management is emerging, and we
+know less about how that will look in the long run, so we think we will need to version those.
+
+The philosophy has to be that we only add fields and do not significantly change the meaning of existing ones, this
+means the messages stay `v1`, but major changes will require bumps. So all message types includes a single digit version.
```

Comment:
why specify 'digit' really all digits following the v prior to the next `.` is part of the version number

Reply:
because it's 1 digit, i dont really want to start putting semver in there :)

Reply:
didn't mean semver. Just that v12 is very possible.

Reply:
that's fair, though if we ever get > 3 versions we screwed up massively haha